onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -450,40 +450,56 @@ static Status PartitionOrtFormatModelImpl(Graph& graph, FuncManager& func_mgr,
 
   std::vector<NodeComputeInfo> node_compute_funcs;
   node_compute_funcs.reserve(nodes_and_viewers.size());
-
-  ORT_RETURN_IF_ERROR(current_ep.Compile(nodes_and_viewers, node_compute_funcs));
-
-  if (node_compute_funcs.size() != nodes_and_viewers.size()) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, type, " did not return correct number of compiled functions");
+  std::unordered_set<size_t> compile_error_indices;
+  compile_error_indices.reserve(nodes_and_viewers.size());
+
+  // We will compile the partitions one by one, and put the result back into node_compute_funcs
+  // In case one of the partitions fails the compile, we can fall back this particular partition to next EP
+  for (size_t i = 0; i < nodes_and_viewers.size(); ++i) {
+    std::vector<IExecutionProvider::FusedNodeAndGraph> single_node_and_viewer{nodes_and_viewers[i]};
+    std::vector<NodeComputeInfo> single_node_compute_func;
+    auto status = current_ep.Compile(single_node_and_viewer, single_node_compute_func);
+    if (!status.IsOK()) {
+      LOGS_DEFAULT(ERROR) << "EP: " << current_ep.Type() << " has Compile error: " << status.ErrorMessage();
+      compile_error_indices.insert(i);
+      node_compute_funcs.push_back(NodeComputeInfo());  // insert an empty stub here
+    } else {
+      ORT_RETURN_IF(single_node_compute_func.empty(), "single_node_compute_func should have 1 elements");
+      node_compute_funcs.push_back(single_node_compute_func[0]);
+    }
   }
 
   for (size_t j = 0, end = nodes_and_viewers.size(); j < end; j++) {
     Node& node = nodes_and_viewers[j].fused_node;
```

Comment:
Why have a separate loop here which requires the insertion of an empty stub above vs. a single loop that does Compile + CancelFuseSubGraph/FinalizeFuseSubGraph etc. in each loop iteration

Reply:
Yes, we only need one loop